src/persist-client/examples/maelstrom/txn.rs

```diff
@@ -0,0 +1,564 @@
+use crate::maelstrom::Args;
+
+pub fn run(args: Args) -> Result<(), Box<dyn std::error::Error>> {
+    let read = std::io::stdin();
+    let write = std::io::stdout();
+
+    crate::maelstrom::node::run::<_, _, TransactorService>(args, read.lock(), write)?;
+    Ok(())
+}
+
+/// Key of the persist shard used by [Transactor]
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct MaelstromKey(u64);
+
+/// Val of the persist shard used by [Transactor]
+#[derive(Debug, Clone, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub struct MaelstromVal(Vec<u64>);
+
+/// An implementation of read-write transactions on top of persist
+///
+/// This executes Maelstrom [txn-list-append] transactions. The timestamp
+/// selection is modeled after Materialize's SQL implementation.
+///
+/// [txn-list-append]: https://github.com/jepsen-io/maelstrom/blob/v0.2.1/doc/workloads.md#workload-txn-list-append
+///
+/// A persist shard is maintained that directly represents a key-value map as it
+/// evolves over time. (Our real SQL implementation would likely instead
+/// maintain a WAL of updates to multiple tables.) Each transaction has
+/// associated timestamps:
+///
+/// - `read_ts`: at which the contents of the map are read (inclusive)
+/// - `write_ts`: at which txn mutations (if any) are written
+/// - `expected_upper`: the upper bound of the previous txn
+/// - `new_upper`: the upper bound after this transaction commits, if it does
+///
+/// To keep things simple, `write_ts` is always `read_ts+1`, `expected_upper` is
+/// `antichain[write_ts]`, and `new_upper` is `antichain[write_ts+1]`.
+///
+/// Transactions are "committed" by using `[WriteHandle::compare_and_append]`,
+/// which atomically:
+/// - verifies that `expected_upper` matches
+/// - writes the updates
+/// - advances the upper to `new_upper`.
+///
+/// This guarantees that all transactions are linearized with each txn's
+/// `read_ts` equal to the previous txn's `write_ts`. If two transactions race
+/// by reading at the same `read_ts` and writing at the same `write_ts`, the
+/// `compare_and_append` guarantees that only one of them succeeds and the other
+/// must retry with new timestamps.
+///
+/// Upon first use, the persist shard is initialized by advancing the upper to
+/// `antichain[T::minimum() + 1]`. This allows the first txn to use 0 as the
+/// `read_ts` and 1 as the `write_ts`.
+///
+/// Similarly modeling Materialize, the since of the shard is kept following the
+/// upper. To keep things simple, this is done by a fixed offset. This exercises
+/// persist compaction.
+///
+/// To ensure that both [ReadHandle::snapshot] and [ReadHandle::listen] are
+/// exercised, when a txn reads the state at `read_ts`, it artificially picks an
+/// `as_of` timestamp in `[since, read_ts]` and splits up the read data between
+/// snapshot and listen along this timestamp.
+#[derive(Debug)]
+pub struct Transactor {
```

Comment:
cc @mjibson. nothing for you to do here necessarily, but possibly relevant to your interests!